packages/compass-data-modeling/src/services/export-diagram.tsx

```diff
@@ -0,0 +1,184 @@
+type DiagramInstance = ReturnType<typeof useDiagram>;
+function mapNodeToDiagramNode(
+  node: ReturnType<DiagramInstance['getNodes']>[number]
+): NodeProps {
+  const { data, type, ...restOfNode } = node;
+  return {
+    ...restOfNode,
+    ...(data as any), // TODO: Type data (or expose these methods from the diagramming package)
+    type: type as NodeType,
+  };
+}
+function mapEdgeToDiagramEdge(
+  edge: ReturnType<DiagramInstance['getEdges']>[number]
+): EdgeProps {
+  const { markerStart, markerEnd, ...restOfEdge } = edge;
+
+  // The diagramming package only allows string based markers
+  if (typeof markerStart !== 'string' || typeof markerEnd !== 'string') {
+    throw new Error('Unexpected edge with non-string markers');
+  }
+
+  return {
+    ...restOfEdge,
+    markerEnd: markerEnd.replace('end-', '') as Marker,
+    markerStart: markerStart.replace('start-', '') as Marker,
+  };
+}
+
+function moveSvgDefsToViewportElement(
+  container: Element,
+  targetElement: Element
+) {
+  const markerDef = container.querySelector('svg defs');
+  if (!markerDef) {
+    return;
+  }
+  const diagramSvgElements = targetElement.querySelectorAll('svg');
+  diagramSvgElements.forEach((svg) => {
+    const pathsWithMarkers = svg.querySelectorAll(
+      'path[marker-end], path[marker-start]'
+    );
+    if (pathsWithMarkers.length !== 0) {
+      const clonedDef = markerDef.cloneNode(true) as SVGMarkerElement;
+      svg.insertBefore(clonedDef, svg.firstChild);
+    }
+  });
+  markerDef.remove();
+}
+
+export async function exportToPng(
+  fileName: string,
+  containerRef: React.RefObject<HTMLDivElement>,
+  diagram: DiagramInstance
+) {
+  const container = containerRef.current;
+  if (!container) {
+    throw new Error('Container reference is not set');
+  }
+  const dataUri = await getExportPngDataUri(container, diagram);
+  downloadFile(dataUri, fileName, () => {
+    ReactDOM.unmountComponentAtNode(container);
+  });
+}
+
+export function getExportPngDataUri(
+  container: HTMLDivElement,
+  diagram: DiagramInstance
+): Promise<string> {
+  return new Promise<string>((resolve, reject) => {
+    const nodes = diagram.getNodes();
+    const edges = diagram.getEdges();
+    ReactDOM.render(
+      <DiagramProvider>
+        <Diagram
+          edges={edges.map(mapEdgeToDiagramEdge)}
+          nodes={nodes.map(mapNodeToDiagramNode)}
+          onlyRenderVisibleElements={false}
+        />
+      </DiagramProvider>,
+      container,
+      () => {
+        rafraf(() => {
+          // For export we are selecting react-flow__viewport element,
+          // which contains the export canvas. It excludes diagram
+          // title, minmap, and other UI elements. However, it also
+          // excludes the svg defs that are currently outside of this element.
+          // So, when exporting, we need to include those defs as well so that
+          // edge markers are exported correctly.
```

Comment:
Wouldn't it be easier to hide what we don't need instead of doing this?

Reply:
> Wouldn't it be easier to hide what we don't need instead of doing this?
> Can we make this configurable in the diagramming package?

These custom marker are rendered [here in the package](https://github.com/mongodb-js/diagramming/blob/main/src/components/canvas/canvas.tsx#L156). I am not sure if i follow what you mean by making it configurable?

Reply:
Sorry, comment somehow attached to the wrong place, should've been this part:

```
It excludes diagram title, minmap, and other UI elements
```

We're doing all this to avoid including controls in the diagram screenshot. Can we change diagramming package to allow us to render it with hidden controls? Then you don't need to do anything to target an element that doesn't include everything that you need